File: backtracking/knight_tour.go
package main

import (
	"fmt"
)

func isValid(x, y, N int, board [][]int) bool {
	return x >= 0 && x < N && y >= 0 && y < N && board[x][y] == -1
}

func solveKTUtil(x, y, movei int, board [][]int, xMove, yMove []int, N int) bool {
	if movei == N*N {
		return true
	}

	for k := 0; k < 8; k++ {
		nextX := x + xMove[k]
		nextY := y + yMove[k]
		if isValid(nextX, nextY, N, board) {
			board[nextX][nextY] = movei
			if solveKTUtil(nextX, nextY, movei+1, board, xMove, yMove, N) {
				return true
			}
			board[nextX][nextY] = -1
		}
	}
	return false
}

func solveKT(N int) bool {
	board := make([][]int, N)
	for i := range board {
		board[i] = make([]int, N)
		for j := range board[i] {
			board[i][j] = -1
		}
	}

	xMove := []int{2, 1, -1, -2, -2, -1, 1, 2}
	yMove := []int{1, 2, 2, 1, -1, -2, -2, -1}

	board[0][0] = 0

	if !solveKTUtil(0, 0, 1, board, xMove, yMove, N) {
		fmt.Println("Solution does not exist")
		return false
	} else {
		printSolution(board)
	}
	return true
}

func printSolution(board [][]int) {
	for _, row := range board {
		for _, val := range row {
			fmt.Printf("%3d ", val)
		}
		fmt.Println()
	}
}

func main() {
	solveKT(5)
}

File: backtracking/knight_tour.rs
/**
 * Knight's Tour Algorithm in Rust
 */

fn is_valid(x: i32, y: i32, n: usize, board: &Vec<Vec<i32>>) -> bool {
    x >= 0 && x < n as i32 && y >= 0 && y < n as i32 && board[x as usize][y as usize] == -1
}

fn solve_kt_util(
    x: i32,
    y: i32,
    movei: i32,
    board: &mut Vec<Vec<i32>>,
    x_move: &[i32],
    y_move: &[i32],
    n: usize,
) -> bool {
    if movei == (n * n) as i32 {
        return true;
    }

    for k in 0..8 {
        let next_x = x + x_move[k];
        let next_y = y + y_move[k];

        if is_valid(next_x, next_y, n, board) {
            board[next_x as usize][next_y as usize] = movei;
            if solve_kt_util(next_x, next_y, movei + 1, board, x_move, y_move, n) {
                return true;
            } else {
                board[next_x as usize][next_y as usize] = -1; // backtracking
            }
        }
    }
    false
}

fn solve_kt(n: usize) -> bool {
    let mut board = vec![vec![-1; n]; n];

    let x_move = [2, 1, -1, -2, -2, -1, 1, 2];
    let y_move = [1, 2, 2, 1, -1, -2, -2, -1];

    board[0][0] = 0;

    if !solve_kt_util(0, 0, 1, &mut board, &x_move, &y_move, n) {
        println!("Solution does not exist");
        return false;
    } else {
        print_solution(&board);
    }
    true
}

fn print_solution(board: &Vec<Vec<i32>>) {
    for row in board {
        for val in row {
            print!("{:3} ", val);
        }
        println!();
    }
}

fn main() {
    solve_kt(5);
}
